engine/src/witness/sol/sol_deposits.rs

```diff
@@ -1,204 +1,111 @@
-									);
-								}
-							}
-						}
-					}
+const MAXIMUM_CONCURRNET_RPCS: usize = 16;
+
+/// We track Solana (Sol and SPL-token) deposits by periodically querying the
+/// state of the deposit channel accounts. To ensure that no deposits are missed
+/// upon fetch, we use a helper program for each deposit channel (FetchHistoricalAccount)
+/// that keeps track of the cumulative amount fetched from the corresponding deposit
+/// channel.
+/// Using the deposit channel's balance and the cumulative amount fetched from it we
+/// can reliably track the amount deposited to a deposit channel.
+/// As a reminder, for token deposit channels the account to witness is the derived
+/// associated account from the actual deposit channel provided by the State Chain.
+/// Native asset and tokens are processed together reusing the getMultipleAccounts
+/// for any combination of deposit channels and assets.
+pub async fn get_channel_ingress_amounts<SolRetryRpcClient>(
+	sol_rpc: &SolRetryRpcClient,
+	vault_address: SolAddress,
+	token_pubkey: SolAddress,
+	deposit_channels: BTreeMap<
+		SolAddress,
+		(OpenChannelDetails<Solana>, ChannelTotalIngressed<Solana>),
+	>,
+) -> Result<BTreeMap<SolAddress, ChannelTotalIngressed<Solana>>, anyhow::Error>
+where
+	SolRetryRpcClient: SolRetryRpcApi + Send + Sync + Clone,
+{
+	let deposit_channels = &deposit_channels;
+	stream::iter(
+		deposit_channels
+			.clone() // https://github.com/rust-lang/rust/issues/110338 && https://github.com/rust-lang/rust/issues/126551
+			.into_iter()
+			.map(|(address, (open_channel_details, _current_total_ingressed))| {
+				match open_channel_details.asset {
+					// Native deposit channel
+					Asset::Sol => DepositChannelType::NativeDepositChannel(address),
+					// Token deposit channel. If we ever want to add another
+					// token we should add a match arm here and adequately
+					// pass a corresponding token_pubkey to this function.
+					Asset::SolUsdc => DepositChannelType::TokenDepositChannel(
+						address,
+						derive_associated_token_account(address, token_pubkey)
+							.expect("Failed to derive associated token account")
+							.address,
+						token_pubkey,
+					),
 				}
-				Ok::<_, anyhow::Error>(())
-			}
-		})
-	}
+			}),
+	)
+	.chunks(MAX_MULTIPLE_ACCOUNTS_QUERY / 2)
+	.map(|deposit_channels_chunk| ingress_amounts(sol_rpc, deposit_channels_chunk, vault_address))
+	.buffered(MAXIMUM_CONCURRNET_RPCS)
+	.map_ok(|(ingress_amounts_chunk, slot)| {
+		stream::iter(
+			ingress_amounts_chunk
+				.into_iter()
+				.filter_map(move |(deposit_channel, amount)| {
+					let amount: SolAmount = amount.saturated_into(); // TODO: Change the DeltaBasedIngress to not use the Chains Amount type but
+												 // instead use u256 or u128 or some generic.
+					let (_, current_total_ingressed) =
+						deposit_channels.get(deposit_channel.address()).unwrap();
+					if amount != current_total_ingressed.amount ||
+						slot < current_total_ingressed.block_number
```

Comment:
I imagine this statement is to avoid scenarios where nodes don't submit the amount because consensus has been reached (with other nodes' votes) and then they don't submit their own. However, I'm not sure that covers all cases.
Example in three nodes with an initial balance of 0 and deposit of 100 at slot 5.
Node A witnesses balance of 100 at slot 5.
Node B witnesses balance of 100 at slot 6.
I assume SC reaches consensus of balance 100 at slot 6.
Let's say node C was slow and didn't pull the deposit channel until after consensus is reached. Then node C witnesses balance of 100 at slot 7. It won't submit the balance since `amount == current_total_ingressed.amount` and `slot > current_total_ingressed.block_number`.

I don't think it's a problem since correct consensus has been reached but up to 1/3 of the nodes might end up not submitting the total_ingressed. I don't expect it to happen on normal behavior though, just wanted to double check that it makes sense.

Reply:
Not sure you totally understand the logic here, but yes the situation you describe technically can happen, i.e. someone doesn't vote as consensus was reached. The `if` statement's purpose to decrease the amount of vote data (by filtering out channels whose deposit amount hasn't changed). Authority's that vote, but don't specify a value for a given channel are assumed to be voting that the amount hasn't changed. This allows authority's to only vote if the amount changes, massively decrease the number of extrinsics/votes needed. Lets have a chat tomorrow to clear it up though.

Reply:
I should have specified, my comment was just about the `slot < current_total_ingressed.block_number`. The other part of the `if` statment makes total sense for the reason you're explaining.
If you agree that the situation that I explained can happen and is not problematic then it's all good :+1: